stats/stats_linux.go

```diff
@@ -0,0 +1,157 @@
+package stats
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"strconv"
+	"time"
+)
+
+type Stats struct {
+	numCpus string
+
+	loadAvg1Min  string
+	loadAvg5Min  string
+	loadAvg15Min string
+
+	memTotal     string
+	memFree      string
+	memAvailable string
+}
+
+// Failures here aren't critical, so we just log them.
+func New(ctx context.Context) *Stats {
+	s := new(Stats)
+
+	if err := s.getNumCpus(); err != nil {
+		s.numCpus = "(unknown)"
+		slog.Info("failed to get cpus", "err", err)
+	}
+
+	go func() {
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				if err := s.getLoadAvgInfo(); err != nil {
+					s.loadAvg1Min = "(unknown)"
+					s.loadAvg5Min = "(unknown)"
+					s.loadAvg15Min = "(unknown)"
+					slog.Info("failed to get loadavg", "err", err)
+				}
+
+				if err := s.getMemInfo(); err != nil {
+					s.memTotal = "(unknown)"
+					s.memFree = "(unknown)"
+					s.memAvailable = "(unknown)"
+					slog.Info("failed to get meminfo", "err", err)
+				}
+			case <-ctx.Done():
+			}
+		}
+	}()
+
+	return s
+
+}
+
+func (s *Stats) GetStatsTags() map[string]string {
+	return map[string]string{
+		"_subtrace_linux_num_cpus":              s.numCpus,
+		"_subtrace_linux_loadavg_1_min":         s.loadAvg1Min,
+		"_subtrace_linux_loadavg_5_min":         s.loadAvg5Min,
+		"_subtrace_linux_loadavg_15_min":        s.loadAvg15Min,
+		"_subtrace_linux_meminfo_mem_total":     s.memTotal,
+		"_subtrace_linux_meminfo_mem_free":      s.memFree,
+		"_subtrace_linux_meminfo_mem_available": s.memAvailable,
+	}
+}
+
+func (s *Stats) getNumCpus() error {
+	info, err := os.ReadFile("/proc/cpuinfo")
+	if err != nil {
+		return fmt.Errorf("read /proc/cpuinfo: %w", err)
+	}
+
+	blocks := bytes.Split(info, []byte{'\n', '\n'})
+
+	// Account for newlines at the end of the file
+	if len(blocks[len(blocks)-1]) == 0 {
```

Comment:
make the function signature `getCPUCount() int` and just return `len(blocks)-1`, then do `fmt.Sprintf("%d", s.getCPUCount())` in the caller. the file is guaranteed to always end with a newline.

also nit: this doesn't need to be a method on `Stats`?